xlators/protocol/server/src/server-rpc-fops_v2.c

```diff
@@ -3186,19 +3203,42 @@ int
 server4_lookup_resume(call_frame_t *frame, xlator_t *bound_xl)
 {
     server_state_t *state = NULL;
+    dict_t *xdata = NULL;
 
     state = CALL_STATE(frame);
 
     if (state->resolve.op_ret != 0)
         goto err;
 
-    if (!state->loc.inode)
+    xdata = state->xdata ? dict_ref(state->xdata) : dict_new();
+    if (!state->loc.inode) {
         state->loc.inode = server_inode_new(state->itable, state->loc.gfid);
-    else
+        if (xdata) {
+            int ret = dict_set_int32(xdata, GF_NAMESPACE_KEY, 1);
+            if (ret) {
+                gf_log(THIS->name, GF_LOG_DEBUG,
+                       "dict set (namespace) failed (path: %s), continuing",
+                       state->loc.path);
+            }
+            if (state->loc.path && (state->loc.path[0] == '<')) {
```

Comment:
Why is this necessary?

Reply:
This is critical. 'namespace' is set on a path. and if the path itself is not resolved, we need to resolve full path. Otherwise, the inode linking code would have some issues. Hence added this code to have this feature built to resolve path from posix layer itself (for faster resolution time). Interestingly I didn't hit this case yet in any of the tests.

Reply:
I think we shouldn't allow lookups based only on gfid. Kernel doesn't do that and it's a basic thing to keep any kind of hierarchical structure consistent. Last time I checked, I think the only place where we really need gfid-based lookups is on a graph switch, but even in this case we could reimplement it in a way that gfid lookups are not really needed.

Anyway this is another discussion and something for another patch.

Reply:
Happy to take this conversation forward. IMO, if we always promise on the hierarchy to be present, we can optimize a lot for many layers and also bring more consistency IMO.

Will keep this code block in the PR, and then fix it properly with the followup PR based on this.